### DISCUSSION TITLE
Cross-compiling nvim - issue with generated files during nvim build

### DISCUSSION THREAD

**Discussion by mwurps** [neovim/neovim#24651] [Q&A] [answered]
Hi all,

I am currently trying to cross-compile nvim to a not (yet) supported OS. So far I have been successful in compiling all of nvims build dependencies(libuv, luv, etc.). But unfortunately the main build of the nvim binary is failing for me and I am not sure if there is a solution to my issue.

I had to find out, that nvim uses lua to generate header/source files during the build process. If I specify the lua binary of my target architecture, then the build process fails, because on my host system I can't execute the target lua binary to generate the source files. If I specify the lua binary of my host system, the build also fails because the host lua binary is trying to load libnlua.so (?) with the src/nvim/generators/preload.lua, which as far as I can tell is built during the build process and is therefore only compatible with my target architecture. The build fails, because my host lua binary can't load the shared object file that's been built for my target architecture...

Does anyone who is familiar with the build process of nvim have any suggestions on how to approach this issue or is it just not (easily) possible to cross-compile nvim?

Thanks for any suggestions

**Comment by clason** [comment, marked as answer]
Unfortunately, there's no support from cross-compiling Nvim at all (for pretty much the reason you've noticed).

Your best bet is to build Nvim in a container for the target architecture.

**Reply by mwurps** [reply]
hmm, thanks

**Comment by danielnachun** [comment]
I was able to cross-compile Nvim relatively easily by first building a copy with the host toolchain, and then setting `LUA_CPATH` to point to the host copy of `libnlua0.so` while building with the cross-compile toolchain.  You can see an example of how I did this for the conda-forge feedstock for Nvim here: https://github.com/conda-forge/nvim-feedstock/blob/main/recipe/build.sh.

The steps are essentially as follows:

1. Build all the dependencies of Nvim with both the host and cross-compile toolchains in advance.  These are separate packages in conda-forge.  This is easier than trying to get cross-compilation to work with the bundled dependencies.
2. Build Nvim with the your host toolchain.  You don't have to install it anywhere permanent as it's essentially just a bootstrap copy.
3. Set `LUA_CPATH` to point to where `libnlua0.so` was build with your host toolchain.  In my case, this was in `${SRC_DIR}/build_host/lib/libnlua0.so`.
4. Use `sed` or some other text replacement tool to replace all instances of `$<TARGET_FILE:nvim_bin>` with the path to the copy of Nvim built with your host toolchain in the following CMakeLists: `src/nvim/po/CMakeLIsts.txt`, `test/CMakeLists.txt`, `runtime/CMakeLists.txt`.
5. Build Nvim with your cross-compile toolchain.

I think supporting cross-compilation could be implemented by just replacing instances of `$<TARGET_FILE:nvim_bin>` with a CMake variable like `BOOTSTRAP_NVIM_BIN` that would default to `$<TARGET_FILE:nvim_bin>` unless overridden by the user.  The user should still be responsible for figuring out how to actually build `nvim` with the host and cross-compile toolchains and how to set `LUA_CPATH` properly.  Hopefully that approach would add no maintenance burden to Nvim itself while avoiding the need to modify any code for cross-compilation to work.
